orangecontrib/bioinformatics/annotation/annotate_samples.py

```diff
@@ -0,0 +1,174 @@
+    p_value_fun : callable, optional (defaults: statistics.Binomial().p_value)
+        A function that calculates p-value. It can be either
+        statistics.Binomial().p_value or hypergeom.sf.
+    """
+
+    def __init__(self, p_value_th, p_value_fun=statistics.Binomial().p_value):
+        if p_value_fun == hypergeom.sf:  # because sf accept x-1 instead of x
+            self.p_value_fun = lambda x, n, m, k: p_value_fun(x-1, n, m, k)
+        else:
+            self.p_value_fun = p_value_fun
+
+        self.p_threshold = p_value_th
+
+    @staticmethod
+    def _select_genes(data):
+        """
+        Function selects "over"-expressed attributes for items with Mann-Whitney
+        U test.
+
+        Parameters
+        ----------
+        data : Orange.data.Table
+            Tabular data
+
+        Returns
+        -------
+        :obj:`list`
+            Sets of selected attributes for each cell
+        """
+        if len(data.X) <= 1:
+            return [], []
+        # rank data
+        data_ge_ranked = rankdata(data.X, axis=0)
+
+        # compute U, mu, sigma
+        n = data_ge_ranked.shape[0]
+        n2 = n - 1
+        u = data_ge_ranked - 1
+        mu = n2 / 2
+        sigma = np.zeros(data_ge_ranked.shape[1])
+        for i in range(data_ge_ranked.shape[1]):
+            _, counts = np.unique(data_ge_ranked[:, i], return_counts=True)
+            sigma[i] = np.sqrt(
+                1 * n2 / 12 * ((n + 1) - np.sum((counts ** 3 - counts)) /
+                               (n * (n - 1))))
+
+        # compute z
+        z = (u - mu) / (sigma + 1e-16)
+
+        # gene selection
+        attributes_np = np.array([
+            a.attributes.get("Entrez ID") for a in data.domain.attributes])
+        attributes_sets = [set(attributes_np[row > 1]) - {None} for row in z]
+        return attributes_sets, z
+
+    @staticmethod
+    def _group_marker_attributes(markers):
+        """
+        Function transforms annotations table to dictionary with format
+        {annotation1: [attributes], annotation2: [attributes], ...}
+        """
+        types_dict = defaultdict(set)
+        for m in markers:
+            types_dict[str(m["Cell Type"])].add(int(m["Entrez ID"].value))
+        return types_dict
+
+    def _assign_annotations(self, items_sets, available_annotations):
+        """
+        Function get set of attributes (e.g. genes) that represents each item
+        and attributes for each annotation. It returns the annotations most
+        significant for each cell.
+
+        Parameters
+        ----------
+        items_sets : list of sets
+            Set of most important attributes for each item.
+        available_annotations : Orange.data.Table
+            Available annotations (e.g. cell types)
+
+        Returns
+        -------
+        ndarray
+            Annotation probabilities
+        list
+            Annotation list
+        """
+        N = 20365  # number of all genes for human
```

Comment:
Number of genes should not be hardcoded inside a methods. Possible solution is to store number of genes as constants at the beginning of the file, based on `tax_id` of the data table.